github/resource_github_repository_ruleset.go

```diff
@@ -0,0 +1,405 @@
+package github
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/google/go-github/v53/github"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/helper/validation"
+)
+
+func resourceGithubRepositoryRuleset() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceGithubRepositoryRulesetCreateOrUpdate,
+		Read:   resourceGithubRepositoryRulesetRead,
+		// Update: resourceGithubRepositoryRulesetUpdate, // TODO: Implement update, replacement fine for now
+		Delete: resourceGithubRepositoryRulesetDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+
+		Schema: map[string]*schema.Schema{
+			"repository": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: "Name of the repository to add the ruleset to.",
+			},
+			"owner": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "Owner of the repository. If not provided, the provider's default owner is used.",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true, // TODO: Remove this when updating is implemented
+				Description: "The name of the Ruleset within the repository.",
+			},
+			"target": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true, // TODO: Remove this when updating is implemented
+				Description: "The target of the ruleset. Either branch or tag.",
+				ValidateFunc: validation.StringInSlice([]string{
+					"branch",
+					"tag",
+				}, false),
+			},
+			"enforcement": {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true, // TODO: Remove this when updating is implemented
+				Description: "The enforcement level of the ruleset. One of active, disabled or evaluate. `evaluate` allows admins to test rules before enforcing them. Admins can view insights on the Rule Insights page (`evaluate` is only available with GitHub Enterprise).",
+				ValidateFunc: validation.StringInSlice([]string{
+					"active",
+					"disabled",
+					"evaluate",
+				}, false),
+			},
+			"conditions": {
+				Type:        schema.TypeList,
+				Optional:    true,
+				ForceNew:    true, // TODO: Remove this when updating is implemented
+				MaxItems:    1,
+				Description: "Target branches/tags. Both an inclusion and exclusion list, supporting regexes as well as ALL branches/tags and the default branch",
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						// TODO: Should the default branch + ALL branches/tags have it's own field?
```

Comment:
I could see an argument for having fields for both `default_branch` and `all_branches_and_tags` as boolean values to get around users needing to learn that they are referenced as ``~DEFAULT_BRANCH`` and ``~ALL``. 

Changing this would be trivial if anyone strongly feels they should be their own fields (updating this after this PR is merged wouldn't be difficult either, if the functionality is wanted down the line)